client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssFetcher.java

```diff
@@ -130,23 +136,26 @@ public void fetchAllRssBlocks() throws IOException, InterruptedException {
   @VisibleForTesting
   public void copyFromRssServer() throws IOException {
     // fetch a block
-    final long startFetch = System.currentTimeMillis();
-    CompressedShuffleBlock compressedBlock = shuffleReadClient.readShuffleBlockData();
-    ByteBuffer compressedData = null;
-    if (compressedBlock != null) {
-      compressedData = compressedBlock.getByteBuffer();
+    if (!retry) {
+      final long startFetch = System.currentTimeMillis();
+      compressedBlock = shuffleReadClient.readShuffleBlockData();
+      if (compressedBlock != null) {
+        compressedData = compressedBlock.getByteBuffer();
+      }
+      long fetchDuration = System.currentTimeMillis() - startFetch;
+      readTime += fetchDuration;
     }
-    long fetchDuration = System.currentTimeMillis() - startFetch;
-    readTime += fetchDuration;
 
     // uncompress the block
     if (compressedData != null) {
-      final long startDecompress = System.currentTimeMillis();
-      byte[] uncompressedData = RssShuffleUtils.decompressData(
-        compressedData, compressedBlock.getUncompressLength(), false).array();
-      unCompressionLength += compressedBlock.getUncompressLength();
-      long decompressDuration = System.currentTimeMillis() - startDecompress;
-      decompressTime += decompressDuration;
+      if (!retry) {
+        final long startDecompress = System.currentTimeMillis();
+        uncompressedData = RssShuffleUtils.decompressData(
+          compressedData, compressedBlock.getUncompressLength(), false).array();
+        unCompressionLength += compressedBlock.getUncompressLength();
+        long decompressDuration = System.currentTimeMillis() - startDecompress;
+        decompressTime += decompressDuration;
```

Comment:
retry need sleep.

Reply:
There is already a "sleep".
If it returns in this round, it will be blocked by `merger.waitForResource()`, which reduces the amount of usedMemory.